sqlmesh/utils/metaprogramming.py

```diff
@@ -633,6 +635,40 @@ def print_exception(
     out.write(tb)
 
 
+def _deterministic_repr(obj: t.Any) -> str:
+    """Create a deterministic representation by ensuring consistent ordering before repr().
+
+    For dictionaries, ensures consistent key ordering to prevent non-deterministic
+    serialization that affects fingerprinting. Uses Python's native repr() logic
+    for all formatting to handle edge cases properly.
+
+    Note that this function assumes list/tuple order is significant and therefore does not sort them.
+
+    Args:
+        obj: The object to represent as a string.
+
+    Returns:
+        A deterministic string representation of the object.
+    """
+
+    def _normalize_for_repr(o: t.Any) -> t.Any:
+        if isinstance(o, dict):
+            # Create a new dict with sorted keys, handling mixed types by using string representation for sorting
+            try:
+                sorted_items = sorted(o.items())
+            except TypeError:
```

Comment:
Should we catch the base `Exception` class here instead? I'm worried `TypeError` may be too specific. To me, we should ideally try to normalize and if it fails for whatever reason just fall back to the previous behavior.

Reply:
Yeah was talking to Iaroslav about this too and it seems like we can just always do the fallback behavior